motoko/invoice-canister/src/invoice/ICPLedger.mo

```diff
@@ -0,0 +1,260 @@
+              kind = #InsufficientFunds({balance});
+            })
+          };
+          case (#TxTooOld kind) {
+            let allowed_window_nanos = kind.allowed_window_nanos;
+            #err({
+              message = ?("Error - Tx Too Old. Allowed window of " # Nat64.toText(allowed_window_nanos));
+              kind = #TxTooOld({allowed_window_nanos});
+            })
+          };
+          case (#TxCreatedInFuture) {
+            #err({
+              message = ?"Error - Tx Created In Future";
+              kind = #TxCreatedInFuture;
+            })
+          };
+          case (#TxDuplicate kind) {
+            let duplicate_of = kind.duplicate_of;
+            #err({
+              message = ?("Error - Duplicate transaction. Duplicate of " # Nat64.toText(duplicate_of));
+              kind = #TxDuplicate({duplicate_of});
+            })
+          };
+        };
+      };
+    };
+  };
+
+  type AccountArgs = {
+    // Hex-encoded AccountIdentifier
+    account : Text;
+  };
+  type BalanceResult = Result.Result<BalanceSuccess, BalanceError>;
+
+  type BalanceSuccess = {
+    balance : Nat;
+  };
+  type BalanceError = {
+    message : ?Text;
+    kind : {
+      #InvalidToken;
+      #InvalidAccount;
+      #NotFound;
+      #Other;
+    };
+  };
+  public func balance(args : AccountArgs) : async BalanceResult {
+    switch (Hex.decode(args.account)) {
+      case (#err err) {
+        #err({
+          kind = #InvalidAccount;
+          message = ?"Invalid account";
+        });
+      };
+      case (#ok account) {
+        let balance = await Ledger.account_balance({account = Blob.fromArray(account)});
+        #ok({
+          balance = Nat64.toNat(balance.e8s);
+        });
+      };
+    };
+  };
+
+  public type GetICPAccountIdentifierArgs = {
+    principal : Principal;
+    subaccount : SubAccount;
+  };
+  public func getICPAccountIdentifier(args : GetICPAccountIdentifierArgs) : Blob {
+    A.accountIdentifier(args.principal, args.subaccount);
+  };
+
+  public type ICPVerifyInvoiceArgs = {
+    invoice : T.Invoice;
+    caller : Principal;
+    canisterId : Principal;
+  };
+  public func verifyInvoice(args : ICPVerifyInvoiceArgs) : async T.VerifyInvoiceResult {
+    let i = args.invoice;
+    let destinationResult = U.accountIdentifierToText({
+      accountIdentifier = i.destination;
+      canisterId = ?args.canisterId;
+    });
+    switch destinationResult {
+      case (#err err) {
+        #err({
+          kind = #InvalidAccount;
+          message = ?"Invalid destination account";
+        });
+      };
+      case (#ok destination) {
+        let balanceResult = await balance({account = destination});
+        switch balanceResult {
+          case (#err err) {
+            #err(err);
+          };
+          case (#ok b) {
+            let balance = b.balance;
+            // If balance is less than invoice amount, return error
+            if (balance < i.amount) {
```

Comment:
should this check account for the transfer fee?

Reply:
I'll create a follow-up issue for this